### PR TITLE
Patch of #534 into dev for v2.5.1

### LINKED ISSUE #532: Consider using `collectFile` to optimise aggregating of results of bin statistics steps
### Description of feature

As described by @maxibor here: https://github.com/nf-core/mag/pull/489#issuecomment-1791336502

Where originally we were getting file name clashes of bins sent to (microbes) and skipped (eukaryotes) around DASTool. This was fixed by Maxime by adding correct prefixes with all the information, but it _may_ make sense to use `collectFile` to merge the contents which would avoid such clashes. However there were some things to explore as described in the PR above.

### PR DESCRIPTION
**Pull request by maxibor**
This is a patch for the 2.5.1 release, adressing the following issues (first pointed out in #534):

In `workflow/mag.nf`
- fix: centrifuge_db needs to be in if/else scope
- fix: collectFile in CAT_SUMMARY and QUAST_BINS_SUMMARY to avoid input file name collision

In `subworkflows/gtdbk.nf`
- fix: busco_db needs to be cast to file to call getBaseName() function, 
- failOnMistmatch needs to be False to handle unrefined genomes coming through the ch_bin_metrics(depth) channel

In `subworkflows/depths.nf`
- fix: input file name collision for MAG_DEPTHS_SUMMARY solved by concatenating files with same name in one with .collectFile()

In `modules/local/cat.nf` and `modules/local/cat_summary.nf`
- fix: bin2classification.names.txt needs to be decompressed for .collectFile() in order to solve the input file name collision, and merging is very simplified

In `bin/combine_tables.py` f9c4f84073a9e5eec35454fd568ccd1eaa202ca8
- fix: handle case where genomes in depth files were not processed by GTDB-Tk
- 
Would close #532 and #530

Battle tested with the following command:
```bash
nextflow run /home/maxime_borry/tmp/mag \
	-profile eva,archgen,debug \
	--input /home/maxime_borry/tmp/mag_test/samplesheet_real_data.csv \
	--outdir test_results \
	--igenomes_base "/home/maxime_borry/SDAG_old/04_genomes/" \
	--host_genome GRCh38 \
	--skip_spades \
	--skip_spadeshybrid \
	--skip_krona \
	--skip_prokka \
	--busco_auto_lineage_prok \
	--bin_domain_classification \
	--ancient_dna \
	--binning_map_mode own \
	--refine_bins_dastool \
	--postbinning_input both \
	--busco_db "/r1/people/maxime_borry/02_db/busco_downloads" \
	--run_gunc \
	--gunc_db /r1/people/maxime_borry/02_db/gunc/gunc_db_progenomes2.1.dmnd \
	--gtdb_db /home/maxime_borry/02_db/gtdb/r207/gtdbtk_r207_v2_data.tar.gz \
	--gtdbtk_min_completeness 1 \
	--gtdbtk_max_contamination 20 \
	--cat_db "/home/maxime_borry/02_db/cat/CAT_prepare_20210107/" \
	-dump-channels \
	-resume \
	-with-tower 
```

Screenshot 2023-12-05 at 17 30 18

## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/mag/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/mag _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/local/cat.nf [hunk 1]
```diff
@@ -11,14 +11,14 @@ process CAT {
...
-    path("raw/*.bin2classification.txt.gz")   , emit: tax_classification_taxids
-    path "versions.yml"                       , emit: versions
+    tuple val(meta), path("*.bin2classification.names.txt")    , emit: tax_classification_names
+    path("*.ORF2LCA.names.txt.gz")                             , emit: orf2lca_classification
+    path("raw/*.ORF2LCA.txt.gz")                               , emit: orf2lca
+    path("raw/*.predicted_proteins.faa.gz")                    , emit: faa
+    path("raw/*.predicted_proteins.gff.gz")                    , emit: gff
+    path("raw/*.log")                                          , emit: log
+    path("raw/*.bin2classification.txt.gz")                    , emit: tax_classification_taxids
+    path "versions.yml"                                        , emit: versions
```
Comment (jfy133) [inline modules/local/cat.nf:21]: Do any of the res tof these files also need a meta? e.g. the names? Or are they not used for anything else?

File: workflows/mag.nf [hunk 2]
```diff
@@ -932,8 +937,13 @@ workflow MAG {
             ch_input_for_postbinning_bins_unbins,
             ch_cat_db
         )
+        ch_cat_summary = CAT.out.tax_classification_names
```
Comment (jfy133) [inline workflows/mag.nf:940]:
```suggestion
        // Group all classification results for each sample in a single file
        ch_cat_summary = CAT.out.tax_classification_names
```

File: workflows/mag.nf [hunk 3]
```diff
@@ -932,8 +937,13 @@ workflow MAG {
...
+            .collectFile(keepHeader: true) {
+                    meta, classification ->
+                    ["${meta.id}.txt", classification]
+            }
         CAT_SUMMARY(
```
Comment (jfy133) [inline workflows/mag.nf:945]:
```suggestion
        // Group all classification results for the whole run in a single file
        CAT_SUMMARY(
```

File: subworkflows/local/gtdbtk.nf [hunk 4]
```diff
@@ -52,7 +53,7 @@ workflow GTDBTK {
     ch_filtered_bins = bins
         .transpose()
         .map { meta, bin -> [bin.getName(), bin, meta]}
-        .join(ch_bin_metrics, failOnDuplicate: true, failOnMismatch: true)
+        .join(ch_bin_metrics, failOnDuplicate: true, failOnMismatch: false)
```
Comment (jfy133) [inline subworkflows/local/gtdbtk.nf:56]:
```suggestion
        .join(ch_bin_metrics, failOnDuplicate: true)
```

File: subworkflows/local/gtdbtk.nf [hunk 5]
```diff
@@ -84,11 +85,11 @@ workflow GTDBTK {
...
     GTDBTK_SUMMARY (
         ch_filtered_bins.discarded.map{it[1]}.collect().ifEmpty([]),
-        GTDBTK_CLASSIFYWF.out.summary.collect().ifEmpty([]),
+        GTDBTK_CLASSIFYWF.out.summary.map{it[1]}.collect().ifEmpty([]),
         [],
-        // GTDBTK_CLASSIFYWF.out.filtered.collect().ifEmpty([]),
         []
-        // GTDBTK_CLASSIFYWF.out.failed.collect().ifEmpty([])
+        // GTDBTK_CLASSIFYWF.out.filtered.map{it[1]}.collect().ifEmpty([]),
+        // GTDBTK_CLASSIFYWF.out.failed.map{it[1]}.collect().ifEmpty([])
```
Comment (jfy133) [inline subworkflows/local/gtdbtk.nf:92]:
```suggestion
```

**Review by jfy133 — APPROVED** [review]

**Review by maxibor — COMMENTED** [review]
> Do any of the res tof these files also need a meta? e.g. the names? Or are they not used for anything else? 

None of these are needed for later, hence no meta

File: modules/local/cat.nf (hunk 1 above)
Comment (maxibor) [reply modules/local/cat.nf:21]: None of these files are re-used later, hence no meta for them.

**Comment by jfy133** [comment]
We can ignore the Prettier error, it's a bug in the template rather than us.

**Comment by jfy133** [comment]
@nf-core-bot fix linting

**Comment by jfy133** [comment]
Thanks @maxibor !!

**Pull request merged**